rules/utils/utils.js

```diff
@@ -12,190 +69,430 @@ function getCallingIdentifier(calleeObject) {
+ *
+ * A string is considered a regular expression if it starts and ends with `/`.
+ *
+ * @param {string} The string to check.
+ * @returns {boolean} Whether or not a string resembles a regular expression.
+ */
+function isStringRegexp(string) {
+    return string[0] === '/' && string[string.length - 1] === '/';
+}
+
+/**
+ * Check if a CallExpression node somewhat resembles an Angular component.
+ *
+ * The following are considered Angular components
+ * ```js
+ * app.factory('kittenService', function() {})
+ *     ^^^^^^^
+ * app.factory('kittenService', kittenService)
+ *     ^^^^^^^
+ * app.factory('kittenService', [])
+ *     ^^^^^^^
+ * asyncFn('value', callback)
+ * ^^^^^^^
+ * ```
+ *
+ * @todo FIXME
+ *
+ * @param {Object} node The CallExpression node to check.
+ * @returns {boolean} Whether or not the node somewhat resembles an Angular component.
+ */
+function isAngularComponent(node) {
+    return node.arguments !== undefined &&
+        node.arguments.length === 2 &&
+        isLiteralType(node.arguments[0]) &&
+        (isIdentifierType(node.arguments[1]) ||
+         isFunctionType(node.arguments[1]) ||
+         isArrayType(node.arguments[1]));
+}
+
+/**
+ * Check whether a CallExpression node defines an Angular controller.
+ *
+ * @param {Object} node The CallExpression node to check.
+ * @returns {boolean} Whether or not the node defines an Angular controller.
+ */
+function isAngularControllerDeclaration(node) {
+    return isAngularComponent(node) &&
+        isMemberExpression(node.callee) &&
+        node.callee.property.name === 'controller';
+}
+
+/**
+ * Check whether a CallExpression node defines an Angular filter.
+ *
+ * @param {Object} node The CallExpression node to check.
+ * @returns {boolean} Whether or not the node defines an Angular filter.
+ */
+function isAngularFilterDeclaration(node) {
+    return isAngularComponent(node) &&
+        isMemberExpression(node.callee) &&
+        node.callee.property.name === 'filter';
+}
+
+/**
+ * Check whether a CallExpression node defines an Angular directive.
+ *
+ * @param {Object} node The CallExpression node to check.
+ * @returns {boolean} Whether or not the node defines an Angular directive.
+ */
+function isAngularDirectiveDeclaration(node) {
+    return isAngularComponent(node) &&
+        isMemberExpression(node.callee) &&
+        node.callee.property.name === 'directive';
+}
+
+/**
+ * Check whether a node defines an Angular service.
+ *
+ * The following are considered services
+ * ```js
+ * app.provider('kittenServiceProvider', function() {})
+ *     ^^^^^^^^
+ * app.factory('kittenService', function() {})
+ *     ^^^^^^^
+ * app.service('kittenService', function() {})
+ *     ^^^^^^^
+ * app.constant('KITTENS', function() {})
+ *     ^^^^^^^^
+ * app.value('KITTENS', function() {})
+ *     ^^^^^
+ * ```
+ *
+ * The following are not considered services
+ * ```js
+ * $provide.factory('kittenService', function() {})
```

Comment:
If these are not considered as service, I think this is a mistake. We should update this isAngularServiceDeclaration method in order to support these exta syntaxes.